homeassistant/components/statistics/sensor.py

```diff
@@ -99,26 +122,48 @@ def __init__(self, entity_id, name, sampling_size, max_age, precision):
         self.total = self.min = self.max = None
         self.min_age = self.max_age = None
         self.change = self.average_change = self.change_rate = None
+
+        self._last_scheduled_purge = dt_util.utc_from_timestamp(0)
         self._update_listener = None
+        self._debounced_updater = debounced_updater
+        self._debounced_updater.function = self._async_debounced_update
+
+    async def _async_debounced_update(self):
+        """
+        Debounced call to update entity.
+
+        To process stats just once (in the end) when a burst of samples is received.
+        """
+        await self.async_update_ha_state(True)
 
     async def async_added_to_hass(self):
         """Register callbacks."""
 
-        @callback
-        def async_stats_sensor_state_listener(entity, old_state, new_state):
+        async def async_stats_sensor_state_listener(entity, old_state, new_state):
             """Handle the sensor state changes."""
             self._unit_of_measurement = new_state.attributes.get(
                 ATTR_UNIT_OF_MEASUREMENT
             )
 
             self._add_state_to_queue(new_state)
 
-            self.async_schedule_update_ha_state(True)
+            if self._max_age:
+                # Do not call update if a purge update is scheduled in the short time
```

Comment:
This should not be necessary? 

Purging is only done inside `async_update`, which is only being called via the debouncer. The debouncer is smart enough that if the call is running, it will cancel the call too (after my latest fix). So you shouldn't need this logic.

Reply:
Hi @balloob, thanks for reviewing :)

The debouncer operates in a time window of **50 ms**, and addresses 1 of the 3 problems detected: the burst incoming, but for the other 2 (schedule repeat and periodic overlapping) I'm admitting up to **1 second** error in the max_age of the sensor. These are very different time windows.

If you see the plot with the new behavior, the number of aborted calls to update from incoming samples (the small green area) is small compared to the other aborts, but I'm not sure if removing that the issue will appear again.

After all the tests I did, I think the key to making it work was to **avoid doing sub-second operations** (as the scheduler uses a 1-second ticking clock). On that assumption, if a sample is received in t=0, but next purging is scheduled in t=0.8, the new sample is added BUT update is not called, as it will be in less than a second.

Reply:
But I don't understand, calling the debouncer actually doesn't abort anything because it's actually behaving like a throttle and not a debouncer (I should change the name in another PR).

Reply:
Sorry @balloob, I was AFK for some days...

> calling the debouncer actually doesn't abort anything because it's actually behaving like a throttle and not a debouncer

I suppose you mean that, with a cooldown of 50ms, given a really long burst of samples, like a lot of samples over 500ms:
*  a real debouncer would do only a final call, at the end of the burst.
*  and the throttle would do ~10 calls, once each 50 ms
right?

(I ask you this because I'm not sure if I'm kind of lost in translation, as English is not my best skill :)

If we are in sync, I'll try to explain myself a little better :)

- The bursts I detected when doing the patch were very close in time: usually 2 or 3 samples in a window of < 25 ms. I didn't detect the problem of a continued burst over time, so for that case, there is no difference between the throttle and the debouncer, as both would do only one call, at the end of the cooldown time. So the current (in _dev_) implementation works as expected to solve this particular problem.
  
  In that context,
  > calling the debouncer actually doesn't abort anything
  
  When a burst of three samples comes, the three samples are added, but only the third one calls the update, as it's desired to happen, so I don't follow you there :(

- But there are more problems caused by incoming samples or schedule updates happening in the **sub-second** window (but separated in time with > 5 * cooldown).

- Imposing a debouncer _aka throttler_ **with a cooldown of 1 full second** would probably fusion both kind of problems, without any timestamp logic to decide calls, but as the debouncer is in `immediate=False`, that would impose a much bigger delay for the sensor to update, and I thought 1-second delay is too much and is more a new bug than a fix :)

- That's why the debouncer handles the burst problem, with a very narrow time-window of 50ms, and the sub-second problem is removed with these **rules of "wait for the next scheduled update" or "do not re-schedule the ~same timestamp"**.


Having said all that, I also think that maybe the original problem could be solved in a simplest way, as this one is the result of multiple additions in the quest of removing the core warning. I'm only confident that this particular sum of patches is working ok.

My bet is that the debouncer could be removed from here, and just deal with the timestamp logic, and not the opposite, but I may be wrong. 

I'll try to remove that condition and do the test, to see if it does not raise the core warning.

Reply:
> and the throttle would do ~10 calls, once each 50 ms
right?

This is correct.

And waiting 50ms is also totally fine. 

So if you had a `helper.Debounce(cooldown=50ms, immediate=True)`, you can get this behavior with _just_ the debouncer:

T: 0ms
Sample comes in.
State is updated.

T: 10ms
Sample comes in.
Debouncer is set to update when cooldown is expired.

T: 20ms
Sample comes in

T: 50ms
Cooldown expires
State is updated.
New cooldown is set.

T: 60ms
Sample comes in
Debouncer is set to update when cooldown is expired.

Reply:
Which seems like desired behavior.

Reply:
> Which seems like desired behavior.

Not exactly :), as in your flow example, there are 2 state updates (I would want just one).

For sure that works for a continued flow like the exposed one, but as I've experienced with this sensor, there is no sample at 60ms: bursts come in small packets, with much longer intervals between them.

My proposal was to remove the initial one, following the main objective of decreasing both:
1. The excessive event bus re-scheduling cycle
2. The excessive number of state updates

The core warning is because (1), but (2) is also related. 
Anyway, maybe addressing (2) is out of context here.

I can try the behavior without this conditional (always calling the debouncer with incoming sample), and also switching to immediate=True, if you prefer.

I'll push the changes if everything works ok (and I'll remove the debouncer mocking in tests, as it is not necessary with immediate=True)

Reply:
The thing is that you are only thinking of your use case. For a lot of people, there will only be 1 state change, no bursts. In that case you don't want to slow it down for them. It's also fine to just increase the cooldown to 1 second. That way we just say that we update max once per second, which seems reasonable.

Reply:
> The thing is that you are only thinking of your use case

You're probably right, but it wasn't my intention (in fact, I had a solution for this issue for myself, by setting a fixed scan_interval for these sensors, but I wanted to solve the generic case)

I thought of the debouncer with a small cooldown to impose a small delay, prioritizing too much the bursts, as I was convinced at first that they were the origin of the issue. it seems that I was wrong.

>  It's also fine to just increase the cooldown to 1 second. That way we just say that we update max once per second, which seems reasonable

I did the change and tested it: sadly, **the warnings appear again** :(
(debouncer with 1s, immediate=True, removed this filter of 'Do not call update if a purge update is scheduled in the short time', so calling debounce update for all incoming samples)

As usual, all combinations worked without warnings in the dev machine, and all** of them failed in the RPI)

** Not all: the combination pushed here worked without warnings for almost a week. Now I'm trying a variant **WITHOUT debouncer**, just avoiding listener cancellation if it is scheduled _in the past_. --> It works too, so I'll push that here